### PR TITLE
Add support of a custom material editor per render pipeline in ShaderGraph [2021.1]

### PR DESCRIPTION
**Pull request by alelievr**
### Purpose of this PR

Add the support of CustomEditorForRenderPipeline keyword in ShaderGraph. Note that these changes need this C++ PR to work: https://ono.unity3d.com/unity/unity/pull-request/112067/_/graphics/material-custom-editor-for-renderpipeline.

Here's an example of the generated code in the shader:

```Shader
        CustomEditorForRenderPipeline "Rendering.HighDefinition.LightingShaderGraphGUI" "UnityEngine.Rendering.HighDefinition.HDRenderPipelineAsset"
        CustomEditorForRenderPipeline "ShaderGraph.PBRMasterGUI" "UnityEngine.Rendering.Universal.UniversalRenderPipelineAsset"

```

---
### Testing status

**Manual Tests**: What did you do?
I tested this feature on multiple ShaderGraphs with both HDRP and URP set as target (they both have a custom editor). I then created a material from this graph and switch between HDRP and URP in the graphic settings.
Here's the result:
CrossPipelineMaterials2
As you can see the material UI is correctly updated when switching between RPs.

**Automated Tests**: 
I added two automated test on C++ side for the new Shader API functions.

**Yamato**: (Select your branch):
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Fcustom-editor-for-render-pipeline/.yamato%252Fall-hdrp.yml%2523All_HDRP_CUSTOM-REVISION/4293179/job

### CODE REVIEW

**Review by TomasKiniulis — APPROVED** [review]
Works as expected. Haven't encountered any PR specific issues.
What I've tested:
- Different material types
- Different material types tatgets between HDRP and URP
- Different shader configurations for HDRP and URP targets
- Tested switching between multiple URP and HDRP assets
- Checked all property types
- Tested pipeline switching with updated shadergraph from previous version
- Switching pipeline with multiple shadergraph selected materials

File: com.unity.render-pipelines.high-definition/Editor/Material/Lit/ShaderGraph/HDLitSubTarget.cs [hunk 1]
```diff
@@ -41,7 +41,6 @@ public HDLitData litData
         };
 
         protected override string[] templateMaterialDirectories => passTemplateMaterialDirectories;
-        protected override string customInspector => "Rendering.HighDefinition.LitShaderGraphGUI";
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Material/Lit/ShaderGraph/HDLitSubTarget.cs:44]: be carefult, this one is protected override string customInspector => "Rendering.HighDefinition.LitShaderGraphGUI"; and it isn't "Rendering.HighDefinition.LightingShaderGraphGUI"; .  
Is it equivalent? does it change something? I guess it should have been lighting, but just checking

File: com.unity.render-pipelines.high-definition/Editor/Material/Eye/ShaderGraph/EyeSubTarget.cs [hunk 2]
```diff
@@ -27,7 +27,6 @@ sealed partial class EyeSubTarget : LightingSubTarget, ILegacyTarget, IRequiresD
         };
 
         protected override string[] templateMaterialDirectories => passTemplateMaterialDirectories;
-        protected override string customInspector => "Rendering.HighDefinition.LightingShaderGraphGUI";
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Material/Eye/ShaderGraph/EyeSubTarget.cs:30]: why is there a change in Images/TerrainPaintHoles.png (above, wasn't able to comment :) )

Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/Material/Eye/ShaderGraph/EyeSubTarget.cs:30]: Hum, probably a merge conflict, I'll revert it

File: com.unity.render-pipelines.high-definition/Editor/Material/Lit/ShaderGraph/HDLitSubTarget.cs (hunk 1 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/Material/Lit/ShaderGraph/HDLitSubTarget.cs:44]: Indeed, it should be LitShaderGraphGUI and not the lighting one. Thanks!

**Review by sebastienlagarde — APPROVED** [review]

**Review by elizabeth-legros — APPROVED** [review]

**Review by jenniferd-unity — APPROVED** [review]

**Review by pbbastian — APPROVED** [review]

**Pull request merged**